liblastlog2/src/lastlog2.c

```diff
@@ -0,0 +1,574 @@
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 3, tty, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for tty: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 4, rhost, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for rhost: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 5, pam_service, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for PAM service: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	int step = sqlite3_step (res);
+
+	if (step != SQLITE_DONE) {
+		if (error)
+			if (asprintf (error, "Delete statement did not return SQLITE_DONE: %d",
+				      step) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	sqlite3_finalize(res);
+
+	return 0;
+}
+
+/* Write a new entry. Returns 0 on success, -1 on failure. */
+int
+ll2_write_entry (const char *lastlog2_path, const char *user,
+		 int64_t ll_time, const char *tty, const char *rhost,
+		 const char *pam_service, char **error)
+{
+	sqlite3 *db;
+	int retval;
+
+	if ((db = open_database_rw (lastlog2_path, error)) == NULL)
+		return -1;
+
+	retval = write_entry (db, user, ll_time, tty, rhost, pam_service, error);
+
+	sqlite3_close (db);
+
+	return retval;
+}
+
+/* Write a new entry with updated login time.
+   Returns 0 on success, -1 on failure. */
+int
+ll2_update_login_time (const char *lastlog2_path, const char *user,
+		       int64_t ll_time, char **error)
+{
+	sqlite3 *db;
+	int retval;
+	char *tty;
+	char *rhost;
+	char *pam_service;
+
+	if ((db = open_database_rw (lastlog2_path, error)) == NULL)
+		return -1;
+
+	if (read_entry (db, user, 0, &tty, &rhost, &pam_service, error) != 0) {
+		sqlite3_close (db);
+		return -1;
+	}
+
+	retval = write_entry (db, user, ll_time, tty, rhost, pam_service, error);
+
+	sqlite3_close (db);
+
+	if (tty)
+		free (tty);
+	if (rhost)
+		free (rhost);
+	if (pam_service)
+		free (pam_service);
+
+	return retval;
+}
+
+
+typedef int (*callback_f)(const char *user, int64_t ll_time,
+			  const char *tty, const char *rhost,
+			  const char *pam_service);
+
+static int
+callback (void *cb_func, int argc, char **argv, char **azColName)
+{
+	char *endptr;
+	callback_f print_entry = cb_func;
+
+	if (argc != 5) {
+		fprintf (stderr, "Mangled entry:");
+		for (int i = 0; i < argc; i++)
+			fprintf (stderr, " %s=%s", azColName[i], argv[i] ? argv[i] : "NULL");
+		fprintf (stderr, "\n");
+		exit (EXIT_FAILURE);
```

Comment:
If this happens when called from PAM it will prevents users from logging in, no?

Reply:
If this prevents an user from logging in or not depends on the admin: he decides if errors from this module are fatal or not.
But calling "exit" here is fatal, this needs to be a return with an error value.

Reply:
The callback return -1 now and printing the error to stderr. sqlite3_exec is returning SQLITE_ABORT, which is returned by ll2_read_all too: "SQL error: query aborted". I think that should be enough because this error is more or less an internal error only, which should never happens or happens at once if someone changes the code in the wrong direction.
I have update the docu example to "optional" with the hint that it is up to the admin how this error should be handled (login is possible or not)